### DISCUSSION TITLE
Stateful Set updateStrategy

### DISCUSSION THREAD

**Discussion by a1994sc** [fluxcd/flux2#2506] [Q&A]
Hello, I have statefulset for my gitea instance but whenever I push a new commit I need to delete the `statefulset/gitea` from cluster for FluxCD to work properly... I found [this](https://github.com/fluxcd/flux/issues/3347) on V1 and tried to play with the `.spec.updateStrategy.type` but that didn't seem to work. Any advise?

**Comment by kingdonb** [comment]
@A1994SC What have you tried exactly, and how did it not work?

Setting `spec.updateStrategy.type` to `RollingUpdate` seems to be the correct advice. When Flux applies the change, Kubernetes should update the pods in a rolling strategy. The other strategy is `OnDelete`, which expects that someone will delete the pods in the stateful set in the correct order so that they can be updated safely.

Since I assume there is only one pod, the `RollingUpdate` strategy should be fine.

Can you post your example in a public repo so we can try to see what went wrong?

**Reply by a1994sc** [reply]
[Here](https://github.com/A1994SC/k3s-gitops/blob/main/clusters/san-jose/apps/gitea/deployment.yaml) is my repo where `.spec.updateStrategy.type` is set to `rollingUpdate`.

**Reply by kingdonb** [reply]
What steps have you tried exactly, in what order, and how did it not work?

> `image: gitea/gitea:1.16.3`

Did you update this to a new image somehow? I don't see that in the commit history at all.

Setting the updateStrategy to `RollingUpdate` should have the effect of restarting the pod(s) when an update comes to gitea.

I see this file has an Image Update policy directive in it:

https://github.com/A1994SC/k3s-gitops/blob/main/clusters/san-jose/apps/gitea/deployment.yaml#L26

Can you please elaborate on what is not working? RollingUpdate strategy does not have anything to do with Flux, it's a part of Kubernetes. I'd love to help you if possible but it's not at all clear how, please provide more information.

**Reply by a1994sc** [reply]
I get the following error when I run `flux get all`

```
NAMESPACE       NAME                            READY   MESSAGE                                                                                                                                                                                                                                                 REVISION                                        SUSPENDED 
flux-system     kustomization/flux-system       False   StatefulSet/apps/gitea dry-run failed, reason: Invalid, error: StatefulSet.apps "gitea" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', 'updateStrategy' and 'minReadySeconds' are forbidden     main/cac4544c15562b1784bd0b3523e13c0038dc4ab1   False
```

**Reply by a1994sc** [reply]
I also appreciate the help so far and sorry for the delay/missing info

**Reply by kingdonb** [reply]
Huh. The error message indicates that `updateStrategy` can be updated, are you sure you didn't change any of the fields that weren't allowed to be changed?

You can use the `flux diff kustomization flux-system` to see which fields Flux thinks have changed. See if there are any unexpected changes in there, please and let us know if this fixes the issue for you!

**Reply by a1994sc** [reply]
I get the follow output on when I run the flux diff
```✗ invalid resource path ""```
You can see that I re-added a `statefulset`, in this case being a mariadb server, and I still get the same error of:   
```
StatefulSet/ingress/mariadb dry-run failed, reason: Invalid, error: StatefulSet.apps "mariadb" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', 'updateStrategy' and 'minReadySeconds' are forbidden      main/373a1df79c5dd8dc4fc8924a6877b7e605784ef4
```
[Link](https://github.com/A1994SC/k3s-gitops/blob/b2c34284bab3b1af0f016ebf0db28214c2ca692b/clusters/san-jose/core/database/mariadb-statefulset.yaml) to my commit

**Reply by kingdonb** [reply]
Sorry, `flux diff kustomization flux-system` takes also a `--path` argument which is mandatory, the local path to compare against your named Kustomization that lives in the cluster.

**Comment by a1994sc** [comment]
I made no modifications to the `statefulsets` and every time the
`kustomization` run (every 10 minutes). I will run that command later today
after work. Thank you for the help.

**Comment by vudzero** [comment]
I do have the exact same issue, every time I try to reconcile my repo, even though I don't modify anything about my StatefulSet I get the error:

`Invalid, error: StatefulSet.apps "device-agent" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', 'updateStrategy' and 'minReadySeconds' are forbidden`. My guess is that flux or kustomize do modify the underlying yaml for every reconcile...

I have to manually delete the StatefulSet and then re reconcile for it to go through properly

I do have the option RollingUpdate present

```
apiVersion: apps/v1 # for versions before 1.9.0 use apps/v1beta2
kind: StatefulSet
metadata:
  name: device-agent
  namespace: kcr-app
  labels:
    app: device-agent
spec:
  selector:
    matchLabels:
      app: device-agent
  serviceName: "device-agent"
  replicas: 2 # by default is 1
  minReadySeconds: 10 # by default is 0
  updateStrategy:
    type: RollingUpdate
```
